cpp/generator.cpp

```diff
@@ -0,0 +1,330 @@
+		Engine &engine = Engine::get();
+		if (engine.get_game()) {
+			log::log(MSG(warn) << "close existing game before loading");
+			return;
+		}
+		auto filename = this->get_variable("load_filename").value<std::string>();
+
+		// create an empty game
+		this->regions.clear();
+		engine.start_game(*this);
+		gameio::load(engine.get_game(), filename);
+	});
+	this->add_action(load_action);
+
+	// create a game using this generator
+	options::OptionAction start_action("generate_game", [this]() {
+		this->create();
+	});
+	this->add_action(start_action);
+
+	// stop game
+	options::OptionAction end_action("end_game", [this]() {
+		Engine &engine = Engine::get();
+		engine.end_game();
+	});
+	this->add_action(end_action);
+}
+
+
+std::shared_ptr<GameSpec> Generator::get_spec() const {
+	return this->spec;
+}
+
+std::vector<std::string> Generator::player_names() const {
+	auto &var = this->get_variable("player_names");
+	auto &names = var.value<options::option_list>();
+
+	std::vector<std::string> result;
+	result.push_back("gaia");
+	for (auto &n : names) {
+		result.push_back(n.str_value());
+	}
+	return result;
+}
+
+void Generator::create_regions() {
+
+	// get option settings
+	int seed = this->getv<int>("generation_seed");
+	int size = this->getv<int>("terrain_size");
+	int base_id = this->getv<int>("terrain_base_id");
+	int p_area = this->getv<int>("player_area");
+
+	rng::RNG rng(seed);
+	Region base(size * 16);
+	base.terrain_id = base_id;
+	std::vector<Region> player_regions;
+
+	int player_count = this->player_names().size() - 1;
+	for (int i = 0; i < player_count; ++i) {
+		log::log(MSG(dbg) << "generate player " << i);
+		double angle = static_cast<double>(i) / static_cast<double>(player_count);
+		int ne = size * 11.0 * sin(2 * M_PI * angle);
+		int se = size * 11.0 * cos(2 * M_PI * angle);
+		coord::tile player_tile{ne, se};
+
+		Region player = base.take_tiles(rng, player_tile, p_area);
+		player.terrain_id = 10;
+
+		Region obj_space = player.take_tiles(rng, player.get_center(), p_area / 5);
+		obj_space.owner = i + 1;
+		obj_space.terrain_id = 8;
+
+		Region trees1 = player.take_random(rng, p_area / 10);
+		trees1.terrain_id = 9;
+		trees1.object_id = 349;
+
+		Region trees2 = player.take_random(rng, p_area / 10);
+		trees2.terrain_id = 9;
+		trees2.object_id = 351;
+
+		Region stone = player.take_random(rng, 5);
+		stone.object_id = 102;
+
+		Region gold = player.take_random(rng, 7);
+		gold.object_id = 66;
+
+		Region forage = player.take_random(rng, 6);
+		forage.object_id = 59;
+
+		Region sheep = player.take_random(rng, 4);
+		sheep.owner = obj_space.owner;
+		sheep.object_id = 594;
+
+		player_regions.push_back(player);
+		player_regions.push_back(obj_space);
+		player_regions.push_back(trees1);
+		player_regions.push_back(trees2);
+		player_regions.push_back(stone);
+		player_regions.push_back(gold);
+		player_regions.push_back(forage);
+		player_regions.push_back(sheep);
```

Comment:
What is your plan for terrain generation? Should a script file be used to decide regions and contents? I thought this would be replaced by some python functions in future so I left out some customisation options atm